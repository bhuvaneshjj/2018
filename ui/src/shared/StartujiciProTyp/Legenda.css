.Legenda {
  font-size: 110%;
}

.Legenda-item {
  font-size: 100%;
  margin-left: 0.5em;
}

.Legenda-item-dokonceno {
  background-color: #35841f;
  color: white;
}

.Legenda-item-nedokonceno {
  background-color: #0b3056;
  color: white;
}

.Legenda-item-na-trase {
  background-color: #683102;
  color: white;
}

.Legenda-item-neaktivni {
  background-color: #868686;
  color: black;
}
