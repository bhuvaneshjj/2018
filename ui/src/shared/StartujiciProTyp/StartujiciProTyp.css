.StartujiciProTyp-grid {
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(2.5em, 2.5em));
  grid-gap: 1px;
  max-width: 53em; /* 20 columns, including gaps */
}

.StartujiciProTyp-item {
  width: 2.5em;
  height: 2em;
  display: flex;
  align-items: center;
  justify-content: center;
}
