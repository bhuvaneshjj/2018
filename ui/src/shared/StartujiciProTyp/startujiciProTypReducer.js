import moment from 'moment';
import { AKTUALNI_ROK } from '../../constants';
import { getTypKategorie } from '../../entities/rocniky/rocnikyReducer';

const findStartCislo = (startujici, startCislo) =>
  startujici.find(element => element.startCislo === startCislo);

export const getStartujiciProTyp = ({
  odstartovani,
  rok = AKTUALNI_ROK,
  typ,
  kategorie,
  ucastnici
}) => {
  const results = [];

  ucastnici.allIds.forEach(id => {
    const ucastnik = ucastnici.byIds[id];
    if (ucastnik[rok]) {
      const { prihlaska, vykon } = ucastnik[rok];

      if (odstartovani && vykon && vykon.startCislo) {
        const { cas, dokonceno, kategorie: kategorieId, startCislo } = vykon;
        if (typ === kategorie[kategorieId].typ) {
          const result = { id, dokonceno, startCislo };
          if (cas) {
            result.cas = moment.duration(cas);
          }
          results.push(result);
        }
      } else if (!odstartovani && prihlaska && prihlaska.startCislo) {
        const { kategorie: kategorieId, startCislo } = prihlaska;
        if (typ === kategorie[kategorieId].typ) {
          results.push({ id, startCislo });
        }
      }
    }
  });

  return results.sort((a, b) => a.startCislo - b.startCislo);
};

export const isStartCisloTaken = ({
  odstartovani,
  rok = AKTUALNI_ROK,
  startCislo,
  typ,
  kategorie,
  ucastnici
}) =>
  !!getStartujiciProTyp({ odstartovani, rok, typ, kategorie, ucastnici }).find(
    startujici => startujici.startCislo === startCislo
  );

const populateRange = (start, end) => {
  if (end >= start) {
    return Array(end - start + 1)
      .fill(1)
      .map((val, index) => start + index);
  }
  return Array(start - end + 1)
    .fill(1)
    .map((val, index) => start - index);
};

export const getStartCislaProTyp = ({
  odstartovani,
  rok = AKTUALNI_ROK,
  typ,
  kategorie,
  rocniky,
  ucastnici
}) => {
  const typKategorie = getTypKategorie({ rok, typ, rocniky });
  const { startCisla } = typKategorie;
  if (!startCisla) {
    return [];
  }

  const startujici = getStartujiciProTyp({ odstartovani, rok, typ, kategorie, ucastnici });
  const results = [];
  startCisla.rozsahy.forEach(rozsah => {
    let range = [];

    const parsed = rozsah.match(/(\d+)-(\d+)/);
    if (parsed) {
      range = populateRange(parseInt(parsed[1], 10), parseInt(parsed[2], 10));
    } else {
      range = [parseInt(rozsah, 10)];
    }
    const result = range.map(cislo => findStartCislo(startujici, cislo) || { startCislo: cislo });
    results.push(...result);
  });

  return results;
};
