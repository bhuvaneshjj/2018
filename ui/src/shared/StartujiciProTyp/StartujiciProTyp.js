import React from 'react';
import PropTypes from 'prop-types';
import momentPropTypes from 'react-moment-proptypes';
import { Table } from 'react-bootstrap';
import { dokoncenoStr } from '../../Util';
import './StartujiciProTyp.css';

const Jeden = ({ startCislo, dokonceno, onClick }) => (
  <td className={`StartujiciProTyp-${dokoncenoStr(dokonceno)[0]}`} onClick={onClick}>
    {startCislo}
  </td>
);

Jeden.propTypes = {
  startCislo: PropTypes.number.isRequired,
  dokonceno: PropTypes.bool,
  onClick: PropTypes.func
};

const STARTUJICICH_NA_RADKU = 10;
const generateTable = startujici => {
  const rows = [];
  while (startujici.length > 0) {
    rows.push(startujici.splice(0, STARTUJICICH_NA_RADKU));
  }

  return rows.map((row, index) => (
    <tr key={index}>{row.map(jeden => <Jeden key={jeden.startCislo} {...jeden} />)}</tr>
  ));
};

const StartujiciProTyp = ({ startujici }) => (
  <div>
    <Table className="StartujiciProTyp-table" bordered condensed striped>
      <tbody>{generateTable(startujici)}</tbody>
    </Table>
  </div>
);

StartujiciProTyp.propTypes = {
  startujici: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      startCislo: PropTypes.number.isRequired,
      dokonceno: PropTypes.bool,
      duration: momentPropTypes.momentDurationObj,
      onClick: PropTypes.func
    }).isRequired
  ).isRequired
};

export default StartujiciProTyp;
