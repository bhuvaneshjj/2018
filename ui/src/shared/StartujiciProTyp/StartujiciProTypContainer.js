import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import {
  decorareStartujici,
  getStartCislaProTyp,
  getStartujiciProTyp
} from './startujiciProTypReducer';
import StartujiciProTyp from './StartujiciProTyp';

const mapStateToProps = state => state;

const mergeProps = (stateProps, dispatchProps, ownProps) => {
  const { entities } = stateProps;
  const { jenStartujici, prihlasky, typ, onClick } = ownProps;

  const startujici = jenStartujici
    ? getStartujiciProTyp({ prihlasky, typ, ...entities })
    : getStartCislaProTyp({ prihlasky, typ, ...entities });

  const decorated = jenStartujici
    ? decorareStartujici(startujici, () => true, onClick)
    : decorareStartujici(startujici, jedenStartujici => !jedenStartujici.id, onClick);
  return {
    startujici: decorated
  };
};

const StartujiciProTypContainer = connect(mapStateToProps, null, mergeProps)(StartujiciProTyp);

StartujiciProTypContainer.propTypes = {
  jenStartujici: PropTypes.bool.isRequired,
  prihlasky: PropTypes.bool.isRequired,
  typ: PropTypes.string.isRequired,
  onClick: PropTypes.func
};

export default StartujiciProTypContainer;
