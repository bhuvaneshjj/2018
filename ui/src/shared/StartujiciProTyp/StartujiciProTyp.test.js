import React from 'react';
import PropTypes from 'prop-types';
import renderer from 'react-test-renderer';
import moment from 'moment';
import StartujiciProTyp from './StartujiciProTyp';

const testRenderer = ({ startCislo }) => (
  <div key={startCislo} className="StartujiciProTyp-item">
    {startCislo}
  </div>
);
testRenderer.propTypes = {
  startCislo: PropTypes.number.isRequired
};

it('žádný startující', () => {
  const component = renderer.create(<StartujiciProTyp startujici={[]} renderer={testRenderer} />);
  expect(component.toJSON()).toMatchSnapshot();
});

it('jeden startující', () => {
  const startujici = [
    {
      id: '10',
      startCislo: 7,
      dokonceno: true,
      duration: moment.duration('PT4H15M32.45S')
    }
  ];

  const component = renderer.create(
    <StartujiciProTyp startujici={startujici} renderer={testRenderer} />
  );
  expect(component.toJSON()).toMatchSnapshot();
});

it('třináct startujících', () => {
  const startujici = [
    { id: '0', startCislo: 7, dokonceno: null },
    { id: '1', startCislo: 4, dokonceno: null },
    {
      id: '10',
      startCislo: 16,
      dokonceno: true,
      duration: moment.duration('PT4H15M32.45S')
    },
    { id: '2', startCislo: 25, dokonceno: false },
    {
      id: '3',
      startCislo: 9,
      dokonceno: true,
      duration: moment.duration('PT2H17M29.14S')
    },
    {
      id: '12',
      startCislo: 15,
      dokonceno: true,
      duration: moment.duration('PT3H59M59.01S')
    },
    { id: '9', startCislo: 1, dokonceno: false },
    { id: '29', startCislo: 8, dokonceno: null },
    { id: '5', startCislo: 59, dokonceno: null },
    { id: '11', startCislo: 43, dokonceno: null },
    { id: '18', startCislo: 42, dokonceno: false },
    {
      id: '10',
      startCislo: 33,
      dokonceno: true,
      duration: moment.duration('PT3H30M22.45S')
    },
    { id: '7', startCislo: 21, dokonceno: false },
    {
      id: '13',
      startCislo: 24,
      dokonceno: true,
      duration: moment.duration('PT3H33M14.15S')
    },
    { id: '34', startCislo: 27, dokonceno: null },
    { id: '14', startCislo: 22, dokonceno: null },
    { id: '58', startCislo: 30, dokonceno: null },
    {
      id: '15',
      startCislo: 23,
      dokonceno: true,
      duration: moment.duration('PT3H27M42.38S')
    },
    { id: '59', startCislo: 26, dokonceno: null }
  ];

  const component = renderer.create(
    <StartujiciProTyp startujici={startujici} renderer={testRenderer} />
  );
  expect(component.toJSON()).toMatchSnapshot();
});
