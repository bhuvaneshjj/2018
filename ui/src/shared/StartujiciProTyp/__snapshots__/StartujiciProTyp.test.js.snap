// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jeden startující 1`] = `
<div>
  <table
    className="StartujiciProTyp-table table table-striped table-bordered table-condensed"
  >
    <tbody>
      <tr>
        <td
          className="StartujiciProTyp-dokonceno"
          onClick={[Function]}
        >
          7
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`třináct startujících 1`] = `
<div>
  <table
    className="StartujiciProTyp-table table table-striped table-bordered table-condensed"
  >
    <tbody>
      <tr>
        <td
          className="StartujiciProTyp-na-trase"
          onClick={undefined}
        >
          7
        </td>
        <td
          className="StartujiciProTyp-na-trase"
          onClick={undefined}
        >
          4
        </td>
        <td
          className="StartujiciProTyp-dokonceno"
          onClick={[Function]}
        >
          16
        </td>
        <td
          className="StartujiciProTyp-nedokonceno"
          onClick={undefined}
        >
          25
        </td>
        <td
          className="StartujiciProTyp-dokonceno"
          onClick={[Function]}
        >
          9
        </td>
        <td
          className="StartujiciProTyp-dokonceno"
          onClick={[Function]}
        >
          15
        </td>
        <td
          className="StartujiciProTyp-nedokonceno"
          onClick={undefined}
        >
          1
        </td>
        <td
          className="StartujiciProTyp-na-trase"
          onClick={undefined}
        >
          8
        </td>
        <td
          className="StartujiciProTyp-na-trase"
          onClick={undefined}
        >
          59
        </td>
        <td
          className="StartujiciProTyp-na-trase"
          onClick={undefined}
        >
          43
        </td>
      </tr>
      <tr>
        <td
          className="StartujiciProTyp-nedokonceno"
          onClick={undefined}
        >
          42
        </td>
        <td
          className="StartujiciProTyp-dokonceno"
          onClick={[Function]}
        >
          33
        </td>
        <td
          className="StartujiciProTyp-nedokonceno"
          onClick={undefined}
        >
          21
        </td>
        <td
          className="StartujiciProTyp-dokonceno"
          onClick={[Function]}
        >
          24
        </td>
        <td
          className="StartujiciProTyp-na-trase"
          onClick={undefined}
        >
          25
        </td>
        <td
          className="StartujiciProTyp-na-trase"
          onClick={undefined}
        >
          22
        </td>
        <td
          className="StartujiciProTyp-na-trase"
          onClick={undefined}
        >
          30
        </td>
        <td
          className="StartujiciProTyp-dokonceno"
          onClick={[Function]}
        >
          23
        </td>
        <td
          className="StartujiciProTyp-na-trase"
          onClick={undefined}
        >
          26
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`žádný startující 1`] = `
<div>
  <table
    className="StartujiciProTyp-table table table-striped table-bordered table-condensed"
  >
    <tbody />
  </table>
</div>
`;
